### ISSUE TITLE
preprocessing layer on keras model to accept input string image for tensorflow serving 

### ISSUE THREAD

**Issue by R-Miner** [tensorflow/serving#878]
I am trying to do tensorflow serving with c# client sending a string image tensor. Trying to do all the preprocessing at export. Following is the way how I tried to do it. But I am not getting the required prediction as the model behaves to a python client. 

This is the code snippet:

string_input = tf.placeholder(tf.string,shape=(None,1))
images = tf.image.decode_png(tf.squeeze(string_input),channels=3)
print(images.shape)
images=tf.reshape(images,[256,256,3])
images = tf.image.convert_image_dtype(images, dtype=tf.float32)
images = tf.expand_dims(images, 0)
print(images.shape)

model =Sequential()
model.add(InputLayer(input_tensor = images, input_shape = (256,256,3)))
model.add(Conv2D(32, (3, 3),input_shape = (256, 256, 3),activation = 'relu'))
model.add(MaxPooling2D(pool_size = (2, 2)))
model.add(Flatten())
model.add(Dense(units = 128, activation = 'relu'))
model.add(Dense(units = 2, activation = 'sigmoid'))
model.load_weights("model.h5")
model.summary()

But it looks like load_weights are not appended to the model when I check the model summary. Also its giving me a prediction of .499 and .5 for any image to be classified into two classes. How can I do the preprocessing effectively. Any help appreciated. I spend a whole lot time for this ans is not successful yet.

Any help appreciated.

**Comment by stolarczyk** [comment]
I had a similar problem. Here is my solution.

I defined the preprocessing part in tensorflow (decoding, resizing and casting to float32).
```
string_inp = tf.placeholder(tf.string, shape=(None,))
imgs_map = tf.map_fn(
    tf.image.decode_image,
    string_inp,
    dtype=tf.uint8
)
imgs_map.set_shape((None, None, None, 3))
imgs = tf.image.resize_images(imgs_map, [120, 160])
imgs = tf.reshape(imgs, (-1, 120, 160, 3))
img_float = tf.cast(imgs, dtype=tf.float32) / 255 - 0.5
```

Then I loaded keras model and defined output tensor by calling model object:
``` 
from keras.models import load_model
model = load_model("...")
output = model(img_float)
```
Finally I used `string_inp` and `output` to define the prediction signature.

**Comment by R-Miner** [comment]
@stolarczyk 

I am getting an error:
'Status(StatusCode=InvalidArgument, Detail="slice index 0 of dimension 0 out of bounds.
	 [[Node: map/TensorArrayUnstack/strided_slice = StridedSlice[Index=DT_INT32, T=DT_INT32, begin_mask=0, ellipsis_mask=0, end_mask=0, new_axis_mask=0, 


I am unable to resolve the issue. Wondering whether you encountered such errors. Any help is appreciated.

**Comment by ymodak** [comment]
@R-Miner Just wanted to follow up. Is this still an issue?

**Comment by ymodak** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by viji-123** [comment]
I am facing the same issue. I am first loading the keras model , then i want to take the input as string. Below is the code.

model = load_model('modelname.h5')
image = tf.placeholder(shape=[None], dtype=tf.string)
builder = saved_model_builder.SavedModelBuilder(export_path)

signature = predict_signature_def(inputs={'images': image},
                                  outputs={'scores': model.output})

with K.get_session() as sess:
    builder.add_meta_graph_and_variables(sess=sess,
                                         tags=[tag_constants.SERVING],
                                         signature_def_map={
                                             'predict': signature})
                                         
    builder.save()      



When i try to predict by sending json (string as input) to google cloud model by CLI
gcloud ml-engine predict --model modelname --version version_3 --json-instances test_data.json 

I get the below error:
['{', '  "error": "Prediction failed: Error during model execution: AbortionError(code=StatusCode.INVALID_ARGUMENT, details=\\"Tensor Placeholder_107:0, specified in either feed_devices or fetch_devices was not found in the Graph\\")"', '}']



Please advice.

**Comment by viji-123** [comment]
@stolarczyk 
I tried using your code, still facing the issue as mentioned above. Please help
